aiida/tools/importexport/dbimport/backends/django/__init__.py

```diff
@@ -562,48 +561,57 @@ def import_data_dj(
                     raise exceptions.ImportValidationError(
                         'Node<{}> already has an outgoing {} link'.format(source.uuid, link_type)
                     )
 
-                # If the indegree is `unique` there cannot already be any other incoming links of that type
-                if indegree == 'unique' and (out_id, link['type']) in existing_input_links:
+                # If the outdegree is `unique_pair`,
+                # then the link labels for outgoing links of this type should be unique,
+                # i.e., the source Node may not have a LinkType of current LinkType, going out,
+                # that also has the current Link label, existing already.
+                elif outdegree == 'unique_pair' and \
+                (in_id, link['label'], link['type']) in existing_outgoing_unique_pair:
+                    raise exceptions.ImportValidationError(
+                        'Node<{}> already has an incoming {} link with label "{}"'.format(
+                            target.uuid, link_type, link['label']
+                        )
+                    )
+
+                # If the indegree is `unique` there cannot already be any other incoming links of that type,
+                # i.e., the target Node may not have a LinkType of current LinkType, coming in, existing already.
+                if indegree == 'unique' and (out_id, link['type']) in existing_incoming_unique:
                     raise exceptions.ImportValidationError(
                         'Node<{}> already has an incoming {} link'.format(target.uuid, link_type)
                     )
 
                 # If the indegree is `unique_pair`,
-                # then the link labels for incoming links of this type should be unique
-                elif indegree == 'unique_pair' and (out_id, link['type']) in existing_input_links:
-                    (_, label) = existing_input_links[in_id, link['type']]
-                    if label == link['label']:
-                        raise exceptions.ImportValidationError(
-                            'Node<{}> already has an incoming {} link with label "{}"'.format(
-                                target.uuid, link_type, link['label']
-                            )
+                # then the link labels for incoming links of this type should be unique,
+                # i.e., the target Node may not have a LinkType of current LinkType, coming in
+                # that also has the current Link label, existing already.
+                elif indegree == 'unique_pair' and \
+                (out_id, link['label'], link['type']) in existing_incoming_unique_pair:
+                    raise exceptions.ImportValidationError(
+                        'Node<{}> already has an incoming {} link with label "{}"'.format(
+                            target.uuid, link_type, link['label']
                         )
+                    )
 
                 # New link
                 links_to_store.append(
```

Comment:
The link checking itself now seems correct, very nice. There is just one thing we realized. In this logic we should not rely on export logic, i.e. what export _can_ legally export. So for example, we should not assume that the contents of an archive are necessarily consistent. This means we have to check that the links _within_ an archive are also consistent. To do this, we simply have to update the `existing_` sets at this point. Then in the next iterations, the added link will be included in the existence check and if the archive itself contains duplicate links, it will be noticed.

Reply:
Done. I have done the same for the SQLAlchemy backend, however, the only way to do this was to not "bulk create" the Links after the for loop, but instead add the new Link in every run using `session.add(Link)`.

Reply:
Btw, also for the SQLAlchemy import, I did a check that the new Links are correctly continuously added and can be found using the QueryBuilder in the subsequent run of the for loop.